Providers/Profile.swift

```diff
@@ -954,8 +1013,24 @@ public class BrowserProfile: Profile {
                 return accumulate(thunks)
             }
 
-            return self.withSyncInputs(nil, function: combined) ??
-                   deferMaybe(synchronizers.map { ($0.0, .NotStarted(.NoAccount)) })
+            let readyDeferred = SyncStateMachine(prefs: self.prefsForSync).toReady(authState)
+            let delegate = profile.getSyncDelegate()
+
+            return readyDeferred >>== self.takeActionsOnEngineStateChanges >>== { ready in
+                function(delegate, self.prefsForSync, ready)
+            }
+        }
+
+        private func syncRemaining(synchronizers: [(EngineIdentifier, SyncFunction)], except statuses: [EngineStatus]) -> Deferred<Maybe<[EngineStatus]>> {
+            let done = Set(statuses.map { $0.0 })
+            let remaining = synchronizers.filter { !done.contains($0.0) }
+            if !remaining.isEmpty {
+                log.info("Just done \(done); now calling syncSeveral \(remaining.map {$0.0})")
+                return syncWith(remaining) >>==  { deferMaybe($0 + statuses) }
```

Comment:
Another way to do this is to have `endSyncingMaybe` collect the result statuses for each sync (line 990); that way this function can just skip the ones that are already done.